crates/engine/tree/src/tree/payload_validator.rs

```diff
@@ -0,0 +1,922 @@
+                return Ok(PayloadValidationOutcome::Invalid { block, error });
+            }
+        };
+
+        debug!(target: "engine::tree", block=?block.num_hash(), ?execution_time, "Block executed");
+
+        // Stop prewarming after execution
+        handle.stop_prewarming_execution();
+
+        // Perform post-execution validation
+        if let Err(consensus_error) = self.consensus.validate_block_post_execution(&block, &output)
+        {
+            trace!(target: "engine::tree", block=?block.num_hash(), ?consensus_error, "Post-execution validation failed");
+            let error = NewPayloadError::Other(Box::new(consensus_error));
+            return Ok(PayloadValidationOutcome::Invalid { block, error });
+        }
+
+        // Compute hashed post state
+        let hashed_state = self.provider.hashed_post_state(&output.state);
+
+        debug!(target: "engine::tree", block=?block.num_hash(), "Calculating block state root");
+
+        debug!(
+            target: "engine::tree",
+            block=?block.num_hash(),
+            ?persisting_kind,
+            run_parallel_state_root,
+            has_ancestors_with_missing_trie_updates,
+            use_state_root_task,
+            config_allows_state_root_task=self.config.use_state_root_task(),
+            "Deciding which state root algorithm to run"
+        );
+
+        let state_root_start = Instant::now();
+        let (state_root, trie_updates) = match self.compute_state_root_with_strategy(
+            &block,
+            &hashed_state,
+            tree_state,
+            persisting_kind,
+            run_parallel_state_root,
+            use_state_root_task,
+            &mut handle,
+            execution_time,
+        ) {
+            Ok(result) => result,
+            Err(error) => return Ok(PayloadValidationOutcome::Invalid { block, error }),
+        };
+
+        let state_root_elapsed = state_root_start.elapsed();
+        self.metrics
+            .block_validation
+            .record_state_root(&trie_updates, state_root_elapsed.as_secs_f64());
+
+        debug!(target: "engine::tree", ?state_root, ?state_root_elapsed, block=?block.num_hash(), "Calculated state root");
+
+        // Ensure state root matches
+        if state_root != block.header().state_root() {
+            // call post-block hook
+            self.on_invalid_block(
+                &parent_header,
+                &block,
+                &output,
+                Some((&trie_updates, state_root)),
+            );
+            let error = NewPayloadError::Other(Box::new(ConsensusError::BodyStateRootDiff(
+                GotExpected { got: state_root, expected: block.header().state_root() }.into(),
+            )));
+            return Ok(PayloadValidationOutcome::Invalid { block, error });
+        }
+
+        // Check if the block is a fork
+        let is_fork =
+            ensure_ok!(self.is_fork(block.sealed_header(), tree_state, canonical_in_memory_state));
```

Comment:
should we determine this here if the tree itself can determine this outside of this impl?

Reply:
great point, removed